{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Colab - Install CITE-pro"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!uv pip install --system \"git+https://github.com/ptggenomics/citepro\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alternatively, the CITE-pro can be installed from our s3 bucket.\n",
    "```\n",
    "!uv pip install --system https://ptgngsdata.s3.us-west-2.amazonaws.com/citepro_pre-release2.zip\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import CITE-pro"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import citepro"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Downloading example dataset\n",
    "\n",
    "For demonstration purposes, you’ll be downloading single-cell multiomic data generated by Proteintech Genomics directly into the Google Colab virtual machine. The data is from human resting PBMCs stained with the MultiPro® Human Discovery Panel (HDP) followed by processing using 10x Genomics Flex chemistry with Feature Barcoding Technology.   \n",
    "\n",
    "It is also possible to upload your own count matrix generated from HDP stained cells. This can be accomplished by mounting a Google Drive containing your pre-processed count matrix.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!curl -LO https://ptgngsdata.s3.us-west-2.amazonaws.com/counts/E44_1_restPBMC_DCpos_filtered_feature_bc_matrix.h5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Additional sample can be downloaded:\n",
    "  *  LPS 4hrs, TI - https://ptgngsdata.s3.us-west-2.amazonaws.com/counts/E44_3_LPS_4h_filtered_feature_bc_matrix.h5\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parameters for create_mudata\n",
    "\n",
    "These are parameters for create_mudata:\n",
    "  * path_count - file path to the count matrix\n",
    "  * allow_file - file that contains barcodes passed filter \\(a.k.a allow list\\)\n",
    "  * sample_id - a sample id string. \\(Optional\\)\n",
    "  * prot_norm - method of protein normalization \\(acceptable values: asinh, clr, none\\)\n",
    "  * celltypist_model - predefined model to predict cell types\n",
    "  * add_3d_umap - whether to generate 3d umap "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Explore the celltypist models\n",
    "\n",
    "Execute the following code to see all the available built-in models from the [Celltypist](https://www.celltypist.org/models)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import celltypist as ct\n",
    "ct.models.models_description()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If any of above model fits your need, use it as the `celltypist_model` parameter, otherwize choose none to omit celltype prediction."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from google.colab import output\n",
    "output.enable_custom_widget_manager()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "create_mu_params  = citepro.nbgui.generate_create_mu_setting()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from google.colab import output\n",
    "output.disable_custom_widget_manager()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Execute the create_mudata\n",
    "\n",
    "Now let's read in this data and create mudata object. This usually takes >10 minutes on colab, but will be faster \\(5-10 minutes\\) on a recent generation of HPC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mudat = citepro.recipe.create_mudata(**create_mu_params)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save the result mudata into .h5mu format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir -p data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "mudat.write_h5mu('data/demo.h5mu')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The MuData can be read back with ```mu.read_h5mu()``` function.\n",
    "\n",
    "\n",
    "```python\n",
    "mudat = mu.read_h5mu('data/demo.h5mu')\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Protein descriptive metadata \n",
    "A few useful antibody descriptive data are already incoporated in the data.\n",
    "  * sum - Sum of all UMIs of this antibody among all cells.\n",
    "  * percent - percent of total UMI count this antibody occupied. Useful for determining whether a antibody needs further titrated. A general rule of thumb is that no antibody should occupy more than 10% of UMI count space, given that this a decently sized \\(>20 antibodies\\)) cocktail. \n",
    "  * median, 75th and 95th - median UMI count across all cells\n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mudat['prot'].var[['gene_ids', 'sum', 'percent','median', '75th', '95th']].sort_values(by='percent', ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Convert MuData to AnnData\n",
    "Let's convert to anndata for the ease of downstrean analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = citepro.recipe.mu_to_ann(mudat)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The converted AnnData object has a structure as below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Basic plotting using scanpy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Feature plots using rna-generated umap, colored by predicted cell types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "sc.pl.embedding(adata, basis= 'X_umap_rna', color = ['celltype_ct_majvote'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Put protein and its encoding transcript expression level side-by-side\n",
    "\n",
    "Please note that the dynamic range of protein are set as follow:\n",
    "  * Minimal - 0.5th percentile of protein level, but levels below 2 will not be displayed.\n",
    "  * Maximal - 99.5th percentile of protein level, but the top of the displayed expression range won't go below 6."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "citepro.nbgui.display_multiomics_feature_plot(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Figures can be downloaded with right click -> save as.."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Scanpy supports generating 3D umap!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.embedding(adata, basis= 'X_umap_rna_3d', color = ['prot:CD8A.65146.1', 'rna:CD8A'], projection='3d')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save the processed AnnData object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write_h5ad('data/demo.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Do not forget to download the saved h5ad file using the colab 'File' tab on the left, as all the data on the colab virtual machine will be lost when disconnected. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
